Framework/PythonInterface/plugins/functions/SpinDiffusion.py

```diff
@@ -0,0 +1,93 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2024 ISIS Rutherford Appleton Laboratory UKRI,
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+# pylint: disable=no-init,invalid-name
+import numpy as np
+from functools import cache
+from numpy import float64
+from numpy.typing import NDArray
+from mantid.api import IFunction1D, FunctionFactory
+from mantid.kernel import IntBoundedValidator
+from mantid.kernel import PhysicalConstants
+from scipy.integrate import quad
+from scipy.special import i0e
+from typing import Tuple
+
+MUON_MAGNETOGYRIC_RATIO = 2 * np.pi * PhysicalConstants.MuonGyromagneticRatio  # MHz / G
+
+
+@cache
+def cached_i0e(d_i: float, t: Tuple[float]) -> NDArray[float64]:
+    """Calculate the zeroth order modified Bessel function, scaled by an exponential term."""
+    return i0e(2 * d_i * np.array(t))
+
+
+def autocorrelation(t: Tuple[float], d_1: float, d_2: float, d_3: float) -> NDArray[float64]:
```

Comment:
Bit unexpected that `quad` will call this with a tuple rather than an array - one could copy to an array here at the top level rather than in `cached_i0e`? Also I think `np.asarray` will only copy if it isn't already an array - but maybe that isn't relevant here.

Reply:
I think the reason I did it this way was because the `functools.cache` decorator was giving an error if the input parameter was a np.array:
```
TypeError: unhashable type: 'numpy.ndarray'
```
So I needed to pass it in as a tuple instead for the cache decorator to accept it. I'm not sure if there might be another way of doing this?